### PR TITLE
Add nonlinear residual pingpong check

### PR DESCRIPTION
**Pull request by arovinelli**
Added the parameter `n_max_nonlinear_pinpong` to `FEproblmeSolve`. this parameter is set by default to 100.
the value of `n_max_nonlinear_pinpong` is used inside `FEProblemBase::checkNonlinearConvergence` as threshold value to determine if pingpong is present and execution of teh current timestep shoudl be interrupted.
The value `n_max_nonlinear_pinpong` can be set from teh executioner block in the input file.

### CODE REVIEW

**Comment by arovinelli** [comment]
@lindsayad and @fdkong this works to catch nonlinear residual pingpong.
The only thing I don't like is the error message I get out of this

```
executioners/nl_pingpong.nl_pingpong: 10 Nonlinear |R| = 3.723309e-11
executioners/nl_pingpong.nl_pingpong: Nonlinear solve did not converge due to DIVERGED_LINE_SEARCH iterations 10
```

but there are not many options to choose from PETSC::SNESConvergedReason

File: framework/src/problems/FEProblemBase.C [hunk 1]
```diff
@@ -6574,12 +6574,34 @@ FEProblemBase::checkNonlinearConvergence(std::string & msg,
...
+    _n_nl_pingpong = 0;
+  }
+  else
+    fnorm_old = system._last_nl_rnorm;
+
+  // ping pong will alwasy start with an increase in the residual value.
+  const bool increase = fnorm > fnorm_old ? true : false;
+
+  if (_n_nl_pingpong == 0 && increase)
+    _n_nl_pingpong += 1;
```
Comment (lindsayad) [inline framework/src/problems/FEProblemBase.C:6595]: Should probably just set this to 1 to be safe

File: framework/src/problems/FEProblemBase.C [hunk 2]
```diff
@@ -6574,12 +6574,34 @@ FEProblemBase::checkNonlinearConvergence(std::string & msg,
...
   if (it == 0)
+  {
     system._initial_residual_after_preset_bcs = fnorm;
+    fnorm_old = fnorm;
+    _n_nl_pingpong = 0;
+  }
+  else
+    fnorm_old = system._last_nl_rnorm;
+
+  // ping pong will alwasy start with an increase in the residual value.
```
Comment (lindsayad) [inline framework/src/problems/FEProblemBase.C:6591]: `alwasy` -> `always`

File: framework/src/problems/FEProblemBase.C [hunk 3]
```diff
@@ -6574,12 +6574,34 @@ FEProblemBase::checkNonlinearConvergence(std::string & msg,
...
+  if (_n_nl_pingpong == 0 && increase)
+    _n_nl_pingpong += 1;
+  else if (_n_nl_pingpong == 0 && !increase)
+    _n_nl_pingpong = 0;
+  else
+  {
+    if ((_n_nl_pingpong % 2 == 1 && !increase) || (_n_nl_pingpong % 2 == 0 && increase))
+      _n_nl_pingpong += 1;
+    else
+      _n_nl_pingpong = 0;
```
Comment (lindsayad) [inline framework/src/problems/FEProblemBase.C:6603]: If I'm not crazy, I think these 4 lines give you all the logic you need

File: framework/src/utils/PetscSupport.C [hunk 4]
```diff
@@ -435,6 +435,10 @@ petscNonlinearConverged(SNES snes,
       *reason = SNES_DIVERGED_LINE_SEARCH;
 #endif
       break;
+
+    case MooseNonlinearConvergenceReason::DIVERGED_NL_RESIDUAL_PINGPONG:
+      *reason = SNES_DIVERGED_LINE_SEARCH;
```
Comment (lindsayad) [inline framework/src/utils/PetscSupport.C:440]: Hmm it would be nice if there was a `SNES_DIVERGED_USER`

Comment (lindsayad) [reply framework/src/utils/PetscSupport.C:440]: But beyond the scope of MOOSE

File: test/tests/executioners/nl_pingpong/nonlinear_residual_pingpong.i [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+[Problem]
+  type = FixedPointProblem
+  fp_tag_name = 'previous'
+[]
+
+[Executioner]
+  type = FixedPointSteady
+  nl_rel_tol = 1e-50
+  nl_abs_tol = 1e-15
+  nl_max_its = 50
```
Comment (lindsayad) [inline test/tests/executioners/nl_pingpong/nonlinear_residual_pingpong.i:67]: Can probably just remove these lines?

File: test/tests/executioners/nl_pingpong/nonlinear_residual_pingpong.i [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+  line_search = none
+  n_max_nonlinear_pinpong = 2
+[]
+
+[Outputs]
+  exodus = true
+[]
```
Comment (lindsayad) [inline test/tests/executioners/nl_pingpong/nonlinear_residual_pingpong.i:74]: Might as well remove output

File: test/tests/executioners/nl_pingpong/tests [hunk 7]
```diff
@@ -0,0 +1,10 @@
+[Tests]
+  [./nl_pingpong]
+    type = RunApp
+    input = 'nonlinear_residual_pingpong.i'
+    expect_out = "Nonlinear solve did not converge due to DIVERGED_LINE_SEARCH"
```
Comment (lindsayad) [inline test/tests/executioners/nl_pingpong/tests:5]: I think I would remove ` due to DIVERGED_LINE_SEARCH` in case sometime in the future we add a `SNES_DIVERGED_USER` or something like that

File: test/tests/executioners/nl_pingpong/tests [hunk 8]
```diff
@@ -0,0 +1,10 @@
...
+    requirement = 'MOOSE shall kill execution when non linear residual pinpong is detected.'
```
Comment (lindsayad) [inline test/tests/executioners/nl_pingpong/tests]: I would reword this as `The system shall consider a nonlinear solve diverged if the nonlinear residual oscillates by a user-controlled number of times.`

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: framework/src/problems/FEProblemBase.C (hunk 1 above)
Comment (lindsayad) [reply framework/src/problems/FEProblemBase.C:6595]: If https://github.com/idaholab/moose/pull/16377#discussion_r534562651 is true, then you can just remove these lines

File: test/tests/executioners/nl_pingpong/nonlinear_residual_pingpong.i (hunk 5 above)
Comment (arovinelli) [reply test/tests/executioners/nl_pingpong/nonlinear_residual_pingpong.i:67]: yes

File: framework/src/problems/FEProblemBase.C (hunk 3 above)
Comment (arovinelli) [reply framework/src/problems/FEProblemBase.C:6603]: yes, you are right.

File: framework/src/problems/FEProblemBase.C (hunk 1 above)
Comment (arovinelli) [reply framework/src/problems/FEProblemBase.C:6595]: I'll remove this as the comment is true

File: test/tests/executioners/nl_pingpong/nonlinear_residual_pingpong.i (hunk 6 above)
Comment (arovinelli) [reply test/tests/executioners/nl_pingpong/nonlinear_residual_pingpong.i:74]: removed

File: test/tests/executioners/nl_pingpong/tests (hunk 8 above)
Comment (arovinelli) [reply test/tests/executioners/nl_pingpong/tests]: reworded as suggested

File: framework/src/utils/PetscSupport.C (hunk 4 above)
Comment (arovinelli) [reply framework/src/utils/PetscSupport.C:440]: agree, somebody should add `SNES_DIVERGED_USER `

**Comment by arovinelli** [comment]
@lindsayad I cannot see why internal apps are failing, but I would be surprised if it is related to this PR

**Comment by loganharbour** [comment]
> I cannot see why internal apps are failing

It's not related. It will be resolved later tonight or tomorrow and we can invalidate then.

File: framework/src/utils/PetscSupport.C (hunk 4 above)
Comment (fdkong) [reply framework/src/utils/PetscSupport.C:440]: I was thinking `SNES_DIVERGED_MAX_IT`    or  `SNES_DIVERGED_LOCAL_MIN` would be better than `SNES_DIVERGED_LINE_SEARCH`

Either way, it is super confusing

**Review by fdkong — CHANGES_REQUESTED** [review]

File: framework/src/utils/PetscSupport.C (hunk 4 above)
Comment (arovinelli) [reply framework/src/utils/PetscSupport.C:440]: for me is the same, it would be nice to have an additional output line letting users know when this is triggered.
But I cannot use std::cout or std::cerr because it will fail prechecks. Don't know what `oss << ` does but even that is masked...

Comment (fdkong) [reply framework/src/utils/PetscSupport.C:440]: Yes, it has nothing to do with `line search `.  `SNES_DIVERGED_LOCAL_MIN` can explain part of secret

File: framework/src/problems/FEProblemBase.C [hunk 9]
```diff
@@ -6574,12 +6574,27 @@ FEProblemBase::checkNonlinearConvergence(std::string & msg,
...
+    fnorm_old = fnorm;
+    _n_nl_pingpong = 0;
+  }
+  else
+    fnorm_old = system._last_nl_rnorm;
+
+  // Check for nonlinear residual pingpong.
+  // Pingpong will always start from a residual increase
+  const bool increase = fnorm > fnorm_old ? true : false;
+  if ((_n_nl_pingpong % 2 == 1 && !increase) || (_n_nl_pingpong % 2 == 0 && increase))
```
Comment (fdkong) [inline framework/src/problems/FEProblemBase.C:6594]: Replace `increase`  with `fnorm > fnorm_old` ?

**Review by fdkong — CHANGES_REQUESTED** [review]
It looks nice for me. The only confusing thing is the converged reason. There is no great solution for that.

I would like to use `SNES_DIVERGED_LOCAL_MIN`  because it could explain part of story. If you live a local min, and then you can not get to the global solution, and your residual has to  ping-pong backward because the solver can not really find a right direction.

File: test/tests/executioners/nl_pingpong/nonlinear_residual_pingpong.i [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+[Problem]
+  type = FixedPointProblem
+  fp_tag_name = 'previous'
+[]
+
+[Executioner]
+  type = FixedPointSteady
+  nl_rel_tol = 1e-50
+  line_search = none
```
Comment (fdkong) [inline test/tests/executioners/nl_pingpong/nonlinear_residual_pingpong.i:66]: We do not really use line search, and so we can not complain on it, haha.

File: framework/src/problems/FEProblemBase.C (hunk 9 above)
Comment (lindsayad) [reply framework/src/problems/FEProblemBase.C:6594]: I agree. Since we now only use `increase` in two spots, I'm in favor of just removing `increase` and directly substituting `fnorm > fnorm_old`. It's more explicit

**Comment by lindsayad** [comment]
> I would like to use `SNES_DIVERGED_LOCAL_MIN` because it could explain part of story. If you live a local min, and then you can not get to the global solution, and your residual has to ping-pong backward because the solver can not really find a right direction.

Let's make this change and remove `increase`, and then this is ready to merge. Most users won't understand what this means, but `SNES_DIVERGED_LINE_SEARCH` is even worse because the line search didn't actually diverge.

**Comment by fdkong** [comment]
> Let's make this change and remove `increase`, and then this is ready to merge. Most users won't understand what `SNES_DIVERGED_LOCAL_MIN` means, but `SNES_DIVERGED_LINE_SEARCH` is even worse because the line search didn't actually diverge.

OK, just added one commit

**Comment by arovinelli** [comment]
> > Let's make this change and remove `increase`, and then this is ready to merge. Most users won't understand what `SNES_DIVERGED_LOCAL_MIN` means, but `SNES_DIVERGED_LINE_SEARCH` is even worse because the line search didn't actually diverge.
> 
> OK, just added one commit

I was pushing it now, but you were faster than me.

**Comment by fdkong** [comment]
> I was pushing it now, but you were faster than me.

It was a competition, haha.

File: framework/src/executioners/FEProblemSolve.C [hunk 11]
```diff
@@ -74,6 +74,11 @@ FEProblemSolve::validParams()
   params.addParam<Real>("nl_div_tol", 1.0e10, "Nonlinear Divergence Tolerance");
   params.addParam<Real>("nl_abs_step_tol", 0., "Nonlinear Absolute step Tolerance");
   params.addParam<Real>("nl_rel_step_tol", 0., "Nonlinear Relative step Tolerance");
+  params.addParam<unsigned int>(
+      "n_max_nonlinear_pingpong",
+      100,
```
Comment (lindsayad) [inline framework/src/executioners/FEProblemSolve.C:79]: I appreciate you making this a large number

**Review by lindsayad — APPROVED** [review]
@fdkong you will need to approve before we can merge (you currently have changes requested and I do not want to dismiss)

**Review by fdkong — APPROVED** [review]
